### DISCUSSION TITLE
How to use 1 button?

### DISCUSSION THREAD

**Discussion by Icesythe7** [davetcc/tcMenu#481] [Q&A] [answered]
I am trying to use the 1 button already on my esp32 (as to not add more unneeded hardware), however I can't figure out how to navigate the menus as the documentation on it seems lacking?
```cpp
    const auto press_type = PollButton();
    static uint8_t last_press_type = 0xFF;

    if (press_type != last_press_type) {
        last_press_type = press_type;
        switch (press_type) {
        case none:
            Serial.println("No Press Detected");
            menuMgr.onMenuSelect(false);
            break;
        case single_press:
            Serial.println("Single Press Detected");
            menuMgr.performDirectionMove(false);
            break;
        case double_press:
            Serial.println("Double Press Detected");
            menuMgr.onMenuSelect(true);
            break;
        case long_press:
            Serial.println("Long Press Detected");
            menuMgr.performDirectionMove(true);
            break;
        default:
            break;
        }
    }
```
This is what I am currently trying given the info I can scrape up but nothing is happening in the menu however the serial output is correct so I know it is being called.

EDIT: To be clear is there not just a simple method to call to advance the menu or select an item etc where we can handle the input on our own and just call a whatever method we need for x input?

**Comment by davetcc** [comment]
So there's a couple of things here, firstly you're better off trying to do something similar to the Up Down button encoder class that's in SwitchInput.cpp/h. This provides the RotaryEncoder interface support that's needed by the menu system these days. It should be fairly simple to implement the above logic in that way. If you look at the 4 way joystick support that will give you ideas how to handle all actions at once within the same class.

**Reply by davetcc** [reply]
See https://github.com/davetcc/IoAbstraction/blob/master/src/SwitchInput.h#L428 and then you'd tell code generator to initialize with no input, and after the `setupMenu()` call you'd set that new encoder you would create to be the rotary encoder with something like `switches.setEncoder(0, myNewEncoderPointer);`

**Reply by davetcc** [reply]
In summary, it will require some degree of understanding of C++ and classes, but if you have that then implementing a class that implements RotaryEncoder that's based on what you see in the above should not be that tricky.

**Reply by Icesythe7** [reply]
I guess I'm not understanding what it wants for a callback or why it is needed if I am handling the button myself and don't want the library to initialize the button pin etc?

```cpp
press_type PollButton() {
    static uint64_t last_press_time = 0;
    static uint64_t first_press_time = 0;
    static bool button_held = false;
    static bool single_press_detected = false;
    static bool double_press_detected = false;
    static bool long_press_detected = false;

    const uint64_t current_time = millis();
    const bool button_state = !(GPIO.in & 1);

    if (button_state && !button_held && (current_time - last_press_time) > debounce_time) {
        // Button pressed
        button_held = true;
        if (single_press_detected) {
            // Double press detected
            single_press_detected = false;
            double_press_detected = true;
            last_press_time = 0; // Reset last press time
            return double_press;
        }
        first_press_time = current_time;
    }

    if (!button_state && button_held) {
        // Button released
        button_held = false;
        last_press_time = current_time;
        if (long_press_detected) {
            long_press_detected = false;
            return none;
        }
        if (double_press_detected) {
            double_press_detected = false;
            return none;
        }
        if (single_press_detected) {
            single_press_detected = false;
            return single_press;
        }
        single_press_detected = true;
    }

    // Check for long press
    if (button_state && button_held && (current_time - first_press_time) > long_press_time) {
        button_held = false;
        single_press_detected = false;
        double_press_detected = false;
        long_press_detected = true;
        last_press_time = 0; // Reset last press time
        return long_press;
    }

    // Reset single press detection if time exceeds double press threshold
    if (!button_state && !button_held && single_press_detected && (current_time - last_press_time) > double_press_time) {
        single_press_detected = false;
        return single_press;
    }

    return none;
}

class EncoderOneButton : public RotaryEncoder, public SwitchListener {
	private:
	    const pinid_t upPin, downPin, backPin, nextPin;
	    SwitchListener* passThroughListener;
	    const bool canRotate;
	public:
        EncoderOneButton(pinid_t pinUp, EncoderCallbackFn callback, uint8_t speed = 20);
        EncoderOneButton(pinid_t pinUp, EncoderListener* listener, uint8_t speed = 20);
	    void onPressed(pinid_t pin, bool held) override;
	    void onReleased(pinid_t pin, bool held) override;

	protected:
	    pinid_t getIncrementPin() const { return intent == SCROLL_THROUGH_SIDEWAYS && canRotate ? nextPin : upPin; }
	    pinid_t getDecrementPin() const { return intent == SCROLL_THROUGH_SIDEWAYS && canRotate ? backPin : downPin; }
	    pinid_t getBackPin() const { return intent == SCROLL_THROUGH_SIDEWAYS && canRotate ? upPin : backPin; }
	    pinid_t getNextPin() const { return intent == SCROLL_THROUGH_SIDEWAYS && canRotate ? downPin : nextPin; }
	};

	#define SW_FLAG_PULLUP_LOGIC 0
	#define SW_FLAG_INTERRUPT_DRIVEN 1
	#define SW_FLAG_INTERRUPT_DEBOUNCE 2
	#define SW_FLAG_ENCODER_IS_POLLING 3

	enum SwitchInterruptMode {
	    /** Do no polling, everything is interrupt driven, it doesn't matter if the interrupt is shared over many keys */
	    SWITCHES_NO_POLLING,
	    /** Poll for keys but the rotary encoder is managed by interrupt */
	    SWITCHES_POLL_KEYS_ONLY,
	    /** Poll for everything, there are no interrupts defined in this mode. Halves the switch poll interval.  */
	    SWITCHES_POLL_EVERYTHING
	};
}
    setupMenu();
    auto* enc = new EncoderOneButton(0, what does it want here and why?);
    switches.setEncoder(0, enc);
```

I'm hoping to have some implementation where a single press will advance the menu arrow, a double press will select the current item and a long press would go back (hoping the menu will just loop around so that a decrement menu arrow is not needed), for the morons like myself can we see in the future a world where we have the option to call a simple function to perform these actions so that we can handle any button logic ourselves?
IE 
menuMgr.goUp();
menuMgr.goDown();
menuMgr.goBack();
menuMgr.select();

**Reply by davetcc** [reply]
I _think_ that as long as you're on the latest version, if you just choose the no input plugin in code generator then you can do the following to increment and decrement the encoder by 1 / -1 without needing any new implementation of rotary encoder, this is because a "no-op" one will be created for you:

`switches.getEncoder(0)->increment(direction);`

You can emulate the select button being pressed:

`menuMgr.onMenuSelect(bool held); `

You can emulate back and next:

`void performDirectionMove(bool dirIsBack);`

You can reset the menu, either back to the current level or completely:

`menuMgr.resetMenu(bool completeReset);`

And with that you should have complete control of the menu. That is pretty much how the keyboard support works as well.

**Reply by davetcc** [reply]
I had forgotten that in the last release we created the "no-op" encoder in the menuMgr if no other was present, this should make life a lot easier for you, as you'd just plug those above calls into your code, but make sure you don't call any of those methods until after setupMenu has been called.

**Reply by Icesythe7** [reply]
I tried as you stated but `switches.getEncoder(0)->increment(direction);` `getEncoder` doesn't take any params also if you select "no input" then you don't even get a navigation arrow on the screen.
What i tried 
```cpp
   const auto press_type = PollButton();
   static uint8_t last_press_type = 0xFF;

   if (press_type != last_press_type) {
       last_press_type = press_type;
       switch (press_type) {
       case none:
           Serial.println("No Press Detected");
           break;
       case single_press:
           Serial.println("Single Press Detected");
           switches.getEncoder()->increment(1);
           break;
       case double_press:
           Serial.println("Double Press Detected");
           menuMgr.onMenuSelect(false);
           break;
       case long_press:
           Serial.println("Long Press Detected");
           menuMgr.resetMenu(false);
           break;
       default:
           break;
       }
   }
   ```
   this is called in my loop() after setupMenu() has been called (in setup())
EDIT: using version 4.2.2 of tcmenu from releases

**Reply by davetcc** [reply]
I know this is a bit slow, but manual input control is a bit of an edge case, most use one of the regular methods and then adapt it slightly.

I looked at the code and can't see why it wouldn't work, so I will need to test it myself, I will make a simple one button menu on a board where I can debug, and quite honestly, if I go that far I'll create a plugin for it and release it with this release.

**Reply by Icesythe7** [reply]
Ok I appreciate it! Just makes it nice to be able to use the in-built button on most dev boards nowadays instead of needing additional hardware.

**Reply by davetcc** [reply, marked as answer]
Don't know if you noticed but one button shipped in 4.3? I've tried it and although not ideal, it does work for very simple menu structures.

**Reply by Icesythe7** [reply]
> Don't know if you noticed but one button shipped in 4.3? I've tried it and although not ideal, it does work for very simple menu structures.

Have been using it, works great for simple menus with boards that have buttons built in already, thanks a bunch!
